src/ir.cpp

```diff
@@ -5910,10 +5910,18 @@ static IrInstSrc *ir_gen_array_access(IrBuilderSrc *irb, Scope *scope, AstNode *
     if (array_ref_instruction == irb->codegen->invalid_inst_src)
         return array_ref_instruction;
 
+    // Create an usize-typed result location to hold the subscript value, this
+    // makes it possible for the compiler to infer the subscript expression type
+    // if needed
+    IrInstSrc *usize_type_inst = ir_build_const_type(irb, scope, node, irb->codegen->builtin_types.entry_usize);
+    ResultLocCast *result_loc_cast = ir_build_cast_result_loc(irb, usize_type_inst, no_result_loc());
```

Comment:
Here's the pattern for doing this for fn call args:

```c++
        ResultLoc *no_result = no_result_loc();
        ir_build_reset_result(irb, scope, source_node, no_result);
        ResultLocCast *result_loc_cast = ir_build_cast_result_loc(irb, arg_type, no_result);
```

Note the use of `ir_build_reset_result`. That will be important when, for example, the expression is used in an `inline while` loop. I think if you add a test with such a construct it might reveal a bug in the current implementation.

Reply:
I removed that because `ir_build_cast_result_loc` already calls it on the "derived" result_loc, but I'll add it back if it's needed.
No idea on how to break the current implementation, I tried declaring an array in an inline while but it worked just fine.

Reply:
Oh! I think perhaps the flaw is then in the fn call args example, adding a superfluous reset there. Let me double check.